### ISSUE TITLE
feat: Differentiable Chamfer Distance & Learnable SuperQuadrics

### ISSUE THREAD

**Issue by KhoiDOO** [KhoiDOO/conquer3d#38]
Implement analytical backward CUDA kernels for one-sided and bidirectional Chamfer distance, add learnable SuperQuadrics primitive sets, and provide end-to-end optimization fitting examples.

**Comment by KhoiDOO** [comment]
## Technical Documentation & Architecture Report

### 1. API Usage & Executable Code Snippets
```python
import torch
from conquer3d.ops import chamfer_distance, one_sided_chamfer_distance

# Initialize two CUDA point clouds with autograd tracking
x = torch.randn(5000, 3, device='cuda', requires_grad=True)
y = torch.randn(7000, 3, device='cuda', requires_grad=True)

# 1. Symmetric Bidirectional Chamfer Loss
loss = chamfer_distance(x, y, squared=True)
loss.backward()

print('x grad shape:', x.grad.shape)  # (5000, 3)
print('y grad shape:', y.grad.shape)  # (7000, 3)

# 2. One-sided Chamfer Query with ground-truth reference
x.grad.zero_()
y_gt = y.detach().clone()  # Non-trainable reference target
dists, indices = one_sided_chamfer_distance(x, y_gt, squared=False)
loss_gt = dists.mean()
loss_gt.backward()
assert y_gt.grad is None  # Atomic accumulation bypassed cleanly
```

### 2. Advantages & Core Capabilities
- **Direct Coalesced Query Gradient Writes**: Each CUDA thread owns an entire 3D query point $i \in [0, N-1]$. Gradients $\nabla_{\mathbf{x}_i}$ are written directly to global memory without atomic contention.
- **Hardware-Accelerated Reference Accumulation**: Contributions to reference points $\nabla_{\mathbf{y}_k} = -\sum_{i: j^*(i)=k} \nabla_{\mathbf{x}_i}$ are accumulated via hardware `atomicAdd`.
- **Zero-Allocation Conditional Autograd**: If the reference point cloud is a fixed target (`y.requires_grad == False`), atomic updates are dynamically bypassed via `ctx.needs_input_grad`.
- **Dual Metric Support**: Exact analytical gradients implemented for both squared Euclidean ($L_2^2$) and true Euclidean ($L_2$) distances with division-by-zero clamping.
- **KD-Tree Acceleration**: The forward pass uses Conquer3D's $O(N \log M)$ KD-Tree traversal rather than $O(N \times M)$ brute-force search.

### 3. Disadvantages & Known Trade-offs
- **Point Cloud Extremity Starvation**: Direct coordinate optimization under Chamfer distance can suffer from Voronoi starvation at distant, thin geometric extremities (e.g., fingertips, toes) when the initial point cloud is centered far away. This is resolved by enclosing initialization radius, unsquared metric, or adding Hausdorff penalties.
- **Float32 Precision**: The backward kernel operates in `float32` precision, matching PyTorch's native GPU deep learning representations.

### 4. Empirical Benchmarks & Performance Metrics
- **Optimization Speed**: 10,000 points fitted to the Stanford Armadillo (50,000 vertices) over 300 Adam steps in **0.65 seconds** (~464 steps/second on RTX 40-series GPU).
- **Convergence**: Chamfer loss reduced by **99.72%** (from `0.076343` down to `0.000216`).
- **Numerical Parity**: Analytical gradients match pure PyTorch brute-force `cdist` down to `5.36e-9` (squared) and `4.33e-7` (unsquared).

### 5. Technical Insights & Architectural Decisions
- The nearest-neighbor operator is differentiable almost everywhere by **Danskin's Theorem**. Discrete indices $\partial j^*(i) / \partial x = 0$ a.e., meaning backpropagation only differentiates through the evaluated pairwise coordinates.
- Covered by PR #37.

**Comment by KhoiDOO** [comment]
Implemented and covered by PR #37.
